File: hilt/src/main/java/com/example/hilt/EngineModule.java
package com.example.hilt;

import dagger.Binds;
import dagger.Module;
import dagger.hilt.InstallIn;
import dagger.hilt.android.components.ActivityComponent;
import dagger.hilt.components.SingletonComponent;

/**
 * author : zhangzf
 * date   : 2021/3/13
 * desc   :
 */

@Module
@InstallIn(ActivityComponent.class)
public abstract class EngineModule {
    @bindGasEngine
    @Binds
    public abstract Engine bindGasEngine(GasEngine gasEngine);
    /*@bindElectricEngine
    @Binds
    public abstract Engine bindElectricEngine(ElectricEngine electricEngine);*/
}

File: jetpack/src/main/java/com/example/jetpack/hilt/EngineModule.kt
package com.example.jetpack.hilt

import dagger.Binds
import dagger.Module
import dagger.hilt.InstallIn
import dagger.hilt.android.components.ActivityRetainedComponent

@Module
@InstallIn(ActivityRetainedComponent::class)
abstract class EngineModule {
    @BindGasEngine
    @Binds
    abstract fun bindGasEngine(gasEngine: GasEngine):Engine

    @BindElectricEngine
    @Binds
    abstract fun bindElectricEngine(electricEngine: BindElectricEngine):Engine
}